src/librustc/ty/mod.rs

```diff
@@ -1578,12 +1579,36 @@ impl<'a, 'gcx, 'tcx> AdtDef {
         }
     }
 
-    pub fn destructor(&self) -> Option<DefId> {
-        self.destructor.get()
+    pub fn destructor(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> Option<DefId> {
+        if self.flags.get().intersects(AdtFlags::IS_DTOR_VALID) {
```

Comment:
This is pre-existing though. The correct solution IMO is to query the trait system but fixing all the users is a bit tricky.